Enable compiling with `crossgen2`

dougbu opened the pull request:
- taking over the good work @davidwrighton has done in #31778

davidwrighton committed:
Enable compiling Asp.NET with crossgen2

davidwrighton committed:
Use use linux crossgen2 instead of linux-musl crossgen2 when targetting linux-musl from the build

davidwrighton committed:
Attempt to make musl work

davidwrighton committed:
Fix typo

davidwrighton committed:
Use correct bits to compile take #3

davidwrighton committed:
Add Pgo data

davidwrighton committed:
Fix mistaken at symbol

dougbu commented:
So far, have only rebased on latest 'main'. Working on test runs using the `crossgen2`ed assemblies.

dougbu commented on src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:
```diff
@@ -125,6 +133,11 @@ This package is an internal implementation of the .NET Core SDK and is not meant
+    <Reference Include="Microsoft.NETCore.App.Crossgen2.$(BuildOsName)-$(BuildArchitecture)"
+        ExcludeAssets="All"
+        PrivateAssets="All"
+        GeneratePathProperty="true" />
```
@davidwrighton, questions from #31778:
> I suggest `Microsoft.NETCore.App.Crossgen2.*` packages should contain a same-named `.props` file that sets a property for the tool path. This didn't make sense for `Microsoft.NETCore.App.Runtime.*` packages but definitely does now. Wouldn't need the path property nor some of the later complications finding the tool if package were more self-describing.
>
> Any possibility of removing the need for workarounds (hacks) like https://github.com/dotnet/aspnetcore/pull/31778/files#diff-eb8fd31a7971d712606a187ea074d0bd5d6ea85aae5a707107fd998216986ea5R117, https://github.com/dotnet/aspnetcore/pull/31778/files#diff-eb8fd31a7971d712606a187ea074d0bd5d6ea85aae5a707107fd998216986ea5R365-R368, and related bits in the future❔

dougbu commented:
@davidwrighton @agocke @Lxiamail finally seeing the expected `BadImageFormatException`s now❕ May be close to done w/ testing.

dougbu commented:
@sebastienros heads up. When I merge this (after removing the last two commits) tomorrow (again, I hope), may see an impact on the perf results. @davidwrighton suspects the initial gains will be modest but we should expect more over time. Let us know if perf _decreases_.

dougbu commented:
@davidwrighton help please. I'm seeing test failures I didn't expect e.g. at https://dev.azure.com/dnceng/public/_build/results?buildId=1138686&view=ms.vss-test-web.build-test-results-tab&runId=34540072&resultId=123944&paneView=debug
``` text
Assert.Throws() Failure
Expected: typeof(System.InvalidOperationException)
Actual:   typeof(System.IO.FileLoadException): Could not load file or assembly 'Microsoft.Extensions.FileProviders.Embedded, Version=42.42.42.42, Culture=neutral, PublicKeyToken=adb9793829ddae60' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
---- System.IO.FileLoadException : Could not load file or assembly 'Microsoft.Extensions.FileProviders.Embedded, Version=42.42.42.42, Culture=neutral, PublicKeyToken=adb9793829ddae60' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
```
In this case, build was on BuildPool.Server.Amd64.VS2019.Open and test ran on a Windows.10.Amd64.Server20H2.Open Helix agent.

davidwrighton commented:
@dougbu, I think those test failures are an issue caused by running on the wrong runtime.

Note the result for https://dev.azure.com/dnceng/public/_build/results?buildId=1138686&view=ms.vss-test-web.build-test-results-tab&runId=34540072&resultId=123876&paneView=debug which actually passes.

I can see in AzDo, that the failing test ran as part of workitem `{ "HelixJobId": "91361424-0db7-42aa-ba2d-5656f83b8c40", "HelixWorkItemName": "Microsoft.Extensions.FileProviders.Embedded.Tests--net461" }` and the successful one ran as part of `{ "HelixJobId": "91361424-0db7-42aa-ba2d-5656f83b8c40", "HelixWorkItemName": "Microsoft.Extensions.FileProviders.Embedded.Tests--net6.0" }`

Looking at those names, I'm suspicious. Does that imply that the failing test actually tried to run the binary on the .NET 4.6.1 runtime, and the succeeding test ran on the .NET 6 runtime. If so, that is an indication that the test passed just fine as we don't expect these images to actually work on .NET 4.6.1 after running crossgen.

dougbu commented on eng/targets/Helix.targets:
```diff
@@ -171,6 +176,21 @@
+        OverwriteReadOnlyFiles="$(OverwriteReadOnlyFiles)"
+        Retries="$(CopyRetryCount)"
+        RetryDelayMilliseconds="$(CopyRetryDelayMilliseconds)"
+        SkipUnchangedFiles="$(SkipCopyUnchangedFiles)" />
```
> Does that imply that the failing test actually tried to run the binary on the .NET 4.6.1 runtime, and the succeeding test ran on the .NET 6 runtime.

Yes, that's almost definitely what it means. I didn't consider the TFM in this block. Should only `<Copy/>` when targeting `net6.0`. I'll add a commit or two to (a) focus temporarily on Windows ➡️ Windows and Linux ➡️ Linux jobs (because the other combinations are failing as consistently as we expect) and (b) fix this targeting issue.

dougbu commented on .azure/pipelines/ci.yml:
```diff
@@ -746,14 +746,14 @@ stages:
-                  /p:CrossgenOutput=false /p:ASPNETCORE_TEST_LOG_DIR=artifacts/log $(_InternalRuntimeDownloadArgs)
+                  /p:ASPNETCORE_TEST_LOG_DIR=artifacts/log $(_InternalRuntimeDownloadArgs)
           displayName: Restore interop projects
         - script: ./eng/build.cmd -ci -nobl -noBuildRepoTasks -noRestore -test -all -noBuildNative -projects eng\helix\helix.proj
```
```suggestion
        - script: ./eng/build.cmd -ci -noBuildRepoTasks -noRestore -test -all -noBuildNative -projects eng\helix\helix.proj
```

dougbu replied:
```suggestion
        - script: ./eng/build.cmd -ci -noBuildRepoTasks -noRestore -test -all -noBuildNative -projects eng\helix\helix.proj
```

dougbu commented:
:ok: @davidwrighton we're down to two distinct problems that I'm not understanding
1. The `Caching_SendFileWithFullContentLength_Cached()` test in src/Servers/HttpSys/test/FunctionalTests/ResponseCachingTests.cs I suspect a consistent speed-up or slowdown of the executing code means the client portion is taking longer than 10 seconds before it issues the second request. (Client must send second request before HTTP cache expires for the test to pass.)
2. Our "Tests: Helix ARM64 matrix" job in the aspnetcore-helix-matrix pipeline is consistently running out of space before getting any useful test results e.g. https://dev.azure.com/dnceng/public/_build/results?buildId=1141412&view=logs&j=99ff3a02-281e-524b-bc15-c0daf2fc9715. This likely means assemblies `crossgen2`ed on Ubuntu 18.04 are consistently failing on Redhat 7, Alpine 3.12, and so on i.e. we're unable to store all of the test failure information.

Neither of these problems block the real work because we don't normally `crossgen` or `crossgen2` before submitting Helix jobs. But, could either indicate a real issue in the assemblies we ship❔

dougbu commented:
> a relative speed-up or slowdown of the executing code

Hmm, that's not very clear. I mean either the client code (that in the `using` block) isn't performing as well as it did before or the other tests are running faster, leading to new contention for HTTP connections. @Tratcher what do you think is likely❔

Background: This PR is both using `crossgen2` and hacking things in order to test everything against the `crossgen2`ed assemblies. We don't use `crossgen` or `crossgen2` in Helix tests outside this PR.

Tratcher commented:
Don't fret over Caching_SendFileWithFullContentLength_Cached, those tests have never been the most reliable. You can quarantine it and open an issue.

davidwrighton commented:
@dougbu I can't tell you if `Caching_SendFileWithFullContentLength_Cached` is a real issue. I would rely on your team to understand the test results.

With respect to the arm64 disk space failure, I suspect that you are getting very close to running out of disk space on all architectures, but that the arm64 binaries are larger than on other architectures. (This is expected. Our arm64 compiler produces a larger output than our x64 compiler). I see that you've injected a disk space check into the set of jobs, but that it executes very early. I would like to see a second `Disk Size` check, but after building the shared fx. My guess is that the available disk space will be rather tight, but I don't have a good sense on just how tight. Alternatively, could you run those tests using one of the Helix build machines that has more available hard disk?

dougbu commented:
/azp run

dougbu committed:
Move to Core-Eng build agents
- should resolve problems w/ disk space in "Tests: Helix ARM64 matrix" job
- need `node` in this job for some reason

dougbu commented:
> With respect to the arm64 disk space failure, I suspect that you are getting very close to running out of disk space on all architectures, but that the arm64 binaries are larger than on other architectures.

Resolved this in my "Move to Core-Eng build agents" commit. But, the size difference we're talking about is pretty big. A regular build seems to use 16G of diskspace and disks on the hosted agents tend to have ~22G available on the / disk. We'll see if the switch **Just Works™️**

dougbu commented:
Success 🎉🎉

@davidwrighton I'll clean this up later today or perhaps tomorrow and make it ready for review. On ops rotation today but&hellip;

@Pilchie this work was _supposed_ to happen in P5. Once this is in, I'd like to try out the backport bot and get it into that branch. Tell mode❔

Pilchie commented:
I'm okay with tell-mode, as this is all infrastructure related.

dougbu commented:
/ping reviewers. This is green (🍏📗🟢💚). Thanks @davidwrighton❕

dougbu commented:
/ping reviewers

wtgodbe reviewed (changes requested):
Code changes all look reasonable, but I suggest adding a bunch of comments to `Microsoft.AspNetCore.App.Runtime.csproj` where you made changes - many of the new properties/items/targets are pretty esoteric/non-obvious

dougbu commented:
> many of the new properties/items/targets are pretty esoteric/non-obvious

That's a bit strange since what @davidwrighton did exactly matches what we have for `crossgen`. Would it help to burn the `crossgen` infrastructure w/ fire❔

wtgodbe commented:
What I mean is that somebody reading that file might be confused as to the purpose of the properties you're setting, etc. in this PR. That's probably true for `crossgen` too. Comments explaining their purpose would be useful

dougbu committed:
Remove `crossgen` tool infrastructure
- remove many properties related only to the `crossgen` tool
  - may have missed a few of course

nits:
- `crossgen` -> `crossgen2` in a few comments
- add / expand a few comments to improve clarity

dougbu commented:
:ok: I've removed `crossgen` stuff as much as I had time to handle right now. May be worth going through all of the properties and item groups in Microsoft.AspNetCore.App.Runtime.csproj to make sure all are used elsewhere in the file or by the SDK. But, this at least simplifies the file slightly.

Two main simplifications:
1. No longer need to copy the `crossgen` tool into the runtime packages we produce. @Pilchie should we announce the removal❔
    - @davidwrighton will Microsoft.NETCore.App.Runtime packages change similarly❔ If yes, may need only one announcement😀
2. Only run one `crossgen2` command for both assemblies and symbols instead of the two we needed when using `crosssgen`

Less important but it is nice to require slightly fewer properties and not to need an extra slash at the end of `$(Crossgen2PlatformAssemblyPaths)` (unlike `$(CrossgenPlatformAssemblyPaths)`.

dougbu commented:
> What I mean is that somebody reading that file might be confused as to the purpose of the properties you're setting, etc. in this PR. That's probably true for crossgen too. Comments explaining their purpose would be useful

I think _part_ of what you were seeing was the new properties weren't in the same hunk as the old ones, hiding a few existing comments. Let me know if the current file looks better.

davidwrighton commented:
@dougbu We're hoping to remove crossgen from the runtime packages soonish, but its a slow process moving everyone off of it.

dougbu commented:
/backport to release/6.0-preview5
